tutorials/haskell/snap-web-app/tutorial.md

```diff
@@ -0,0 +1,482 @@
+So, this module is really important as it holds our application stack and resource handlers.
+
+As for the [Main module](https://github.com/stackbuilders/tutorials/tutorials/haskell/snap-web-app/code/src/Todo/Main.hs),
+it just bootstraps the web server using the previous Snaplet created in the
+function `app` and the application configuration. You rarely will need to modify
+it, but we encourage you take a look at it and understand what happens there.
+
+## Setting up the model and migrations
+
+This application goes about managing Todo items, which requires a very simple
+data type definition with fields for description, completion status and,
+of course, a primary key. This is defined in
+[Db module](https://github.com/stackbuilders/tutorials/tutorials/haskell/snap-web-app/code/src/Todo/Db.hs)
+and this is how it looks like using
+[record syntax](https://en.wikibooks.org/wiki/Haskell/More_on_datatypes#Named_Fields_.28Record_Syntax.29):
+
+```haskell
+data Todo = Todo
+  { todoID      :: Maybe Int
+  , description :: Text
+  , completed   :: Bool
+  } deriving (Eq, Show, Read)
+```
+[Back in the Site module](https://github.com/stackbuilders/tutorials/tutorials/haskell/snap-web-app/code/src/Todo/Site.hs#L143),
+there were used a function called `checkMigrationOrMigrate`
+receiving a database connection as parameter, while initalizing our Todo App.
+This function verifies if `todos` table exists in the database and if not, then creates it.
+
+```haskell
+checkMigrationOrMigrate :: Connection
+                        -> IO ()
+checkMigrationOrMigrate conn = do
+  schemaCreated <- tableExists conn "todos"
+  unless schemaCreated $ createTodoTable conn
+```
+
+In order to achieve its purpose, `checkMigrationOrMigrate` relies in a couple of
+aditional functions like `tableExists`:
+
+```haskell
+tableExists :: Connection
+            -> String
+            -> IO Bool
+tableExists conn tblName = do
+  let query = [i| SELECT name
+                  FROM sqlite_master
+                  WHERE type='table' AND name=? |]
+  name <- listToMaybe . fmap rowToValue <$> quickQuery' conn query [toSql tblName]
+  return (name == Just tblName)
+  where
+    rowToValue :: [SqlValue] -> String
+    rowToValue (snippetBody:_) = fromSql snippetBody
+    rowToValue _ = error "Unable to read single value"
+```
+
+And `createTodoTable`:
+
+```haskell
+createTodoTable :: Connection
+                -> IO ()
+createTodoTable conn = do
+  let query = [i| CREATE TABLE todos (
+                  id INTEGER PRIMARY KEY,
+                  description TEXT,
+                  completed BOOLEAN)|]
+
+  quickQuery' conn query []
+  commit conn
+```
+
+The function `tableExists` may look kind of tricky at first sight. It runs a very specific query
+to know if a table exists and then it maps this particular function composition:
+`listToMaybe . fmap rowToValue` over the query result, which should be the table name
```

Comment:
@nebtrx There's something weird with the colon here. What about "It runs a very specific query to know if a table exists and then it maps `listToMaybe . fmap rowToValue`" over the query result (or something else to avoid the colon and then continuing with the same sentence)?